src/devices/MatrixKeyboard/MatrixKeyboard.cs

```diff
@@ -132,31 +111,8 @@ public Task<MatrixKeyboardEventArgs> ReadKeyAsync(CancellationToken token)
         {
             return Task.Run(() =>
             {
-                MatrixKeyboardEventArgs args = null;
-
                 OpenPins();
-                while (!token.IsCancellationRequested)
-                {
-                    for (int i = 0; i < _inputPins.Length; i++)
-                    {
-                        int index = _currentOutput * _inputPins.Length + i;
-
-                        PinValue oldValue = _buttonValues[index];
-                        PinValue newValue = _masterGpioController.Read(_inputPins[i]);
-
-                        _buttonValues[index] = newValue;
-                        if (newValue != oldValue)
-                        {
-                            args = new MatrixKeyboardEventArgs(newValue == PinValue.High ? PinEventTypes.Rising : PinEventTypes.Falling, _currentOutput, i);
-                        }
-                    }
-
-                    _masterGpioController.Write(_outputPins[_currentOutput], PinValue.Low);
-                    _currentOutput = (_currentOutput + 1) % _outputPins.Length;
-                    _masterGpioController.Write(_outputPins[_currentOutput], PinValue.High);
-
-                    Thread.Sleep(ScanInterval);
-                }
+                MatrixKeyboardEventArgs args = Scan(token, true);
```

Comment:
Do all of these functions need to open and close pins? can you just open them in the constructor and close in the dispose method?